packages/tpetra/CMakeLists.txt

```diff
@@ -24,6 +24,173 @@ IF (NOT ${PROJECT_NAME}_ENABLE_CXX11 AND NOT TPL_ENABLE_CUDA)
+
+IF (NOT TPL_ENABLE_CUDA)
+  MESSAGE (STATUS "  - TPL_ENABLE_CUDA is OFF, so we assume that MPI is not CUDA aware.")
+  SET (Tpetra_ASSUME_CUDA_AWARE_MPI_DEFAULT OFF)
+ELSEIF (NOT TPL_ENABLE_MPI)
+  # Even if CUDA is enabled but MPI is not, Teuchos::SerialComm may
+  # assume host access of input and output buffers (e.g.,
+  # Teuchos::reduceAll on a SerialComm may copy input into output on
+  # host).  Thus, it's safer to assume that "MPI" (rather,
+  # Teuchos::SerialComm, a kind of MPI stub) can't access CUDA memory.
+  MESSAGE (STATUS "  - TPL_ENABLE_CUDA is ON but TPL_ENABLE_MPI is OFF, so we assume that (nonexistent) MPI is not CUDA aware.")
+  SET (Tpetra_ASSUME_CUDA_AWARE_MPI_DEFAULT OFF)
+ELSEIF (DEFINED Tpetra_ASSUME_CUDA_AWARE_MPI)
+  # The user set Tpetra_ASSUME_CUDA_AWARE_MPI explicitly.
+  # Just use that value as the default.
+
+  IF (Tpetra_ASSUME_CUDA_AWARE_MPI)
+    SET (Tpetra_ASSUME_CUDA_AWARE_MPI_DEFAULT ON)
+    MESSAGE (STATUS "  - You explicitly set the CMake variable Tpetra_ASSUME_CUDA_AWARE_MPI=ON.  This means that you assume that the MPI implementation you want to use with Trilinos is CUDA aware.  If this is NOT true, then solvers will segfault or otherwise fail.")
+  ELSE ()
+    SET (Tpetra_ASSUME_CUDA_AWARE_MPI_DEFAULT OFF)
+    MESSAGE (STATUS "  - You explicitly set the CMake variable Tpetra_ASSUME_CUDA_AWARE_MPI=OFF.  This means that you assume that the MPI implementation you want to use with Trilinos is NOT CUDA aware.  This is always the safe option, but it may reduce performance if it is not true.")
+  ENDIF ()
+ELSE ()
+  # The user did not set Tpetra_ASSUME_CUDA_AWARE_MPI explicitly.
+
+  # This is annoying, but necessary, because expressions like "DEFINED
+  # X AND (NOT X)" may not be valid CMake syntax.
+  SET (Tpetra_CROSSCOMPILING OFF)
+  IF (DEFINED CMAKE_CROSSCOMPILING)
+    IF (CMAKE_CROSSCOMPILING)
+      SET (Tpetra_CROSSCOMPILING ON)
+    ENDIF ()
+  ENDIF ()
+
+  IF (Tpetra_CROSSCOMPILING)
+    MESSAGE (STATUS "  - Since you are cross compiling, if you wish Tpetra to assume that MPI is CUDA aware, you must set the CMake variable Tpetra_ASSUME_CUDA_AWARE_MPI:BOOL=ON explicitly at configure time.  For now, Tpetra will assume that MPI is NOT CUDA aware; this is always the safe option, but it may reduce performance if it is not true.")
+  ELSE ()
+    # We are not cross compiling, so it is safe to try to auto-detect
+    # by searching for and running the "ompi_info" executable.
+    MESSAGE (STATUS "  - Attempt to detect whether MPI is CUDA aware, by searching for \"ompi_info\" executable.")
+
+    SET (Tpetra_FOUND_OMPI_INFO_EXECUTABLE OFF)
+
+    # First, try the directory where MPI_CXX_COMPILER lives.
+    # This only makes sense if MPI_USE_COMPILER_WRAPPERS is ON.
+
+    IF (DEFINED MPI_USE_COMPILER_WRAPPERS)
+      IF (MPI_USE_COMPILER_WRAPPERS AND DEFINED MPI_CXX_COMPILER AND (NOT Tpetra_FOUND_OMPI_INFO_EXECUTABLE))
```

Comment:
It can't be, but I'm being explicit just to imitate the pattern of the cases below.  I should probably break out each check into a CMake function, but I'm not sure if it would save many lines of code.